### PR TITLE
Added EFI support for plane

### PR DESCRIPTION
**Pull request by tridge**
This allows for logging of electronic fuel injection systems, with support for the MegaSquirt serial interface as first implementation
Also adds a idle governor for ICE

### CODE REVIEW

File: libraries/AP_EFI/AP_EFI.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    // @Description: Used to calibrate fuel flow for MS protocol (Slope)
+    // @Range: 0 1
+    // @User: Advanced
+    // @RebootRequired: False
+    AP_GROUPINFO("_COEF1", 2, AP_EFI, coef1, 0),
+
+    // @Param: _COEF2
+    // @DisplayName: EFI Calibration Coefficient 2
+    // @Description: Used to calibrate fuel flow for MS protocol (Offset)
+    // @Values: 0 10
```
Comment (WickedShell) [inline libraries/AP_EFI/AP_EFI.cpp]: This should have been a range, not values.

File: libraries/AP_EFI/AP_EFI_Serial_MS.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    // value individiually
+    uint16_t message_length = 0;
+
+    // reset checksum before reading new data
+    checksum = 0;
+    
+    // Message length field begins the message (16 bits, excluded from CRC calculation)
+    // Message length value excludes the message length and CRC bytes 
+    message_length = port->read() << 8;
+    message_length += port->read();
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_MS.cpp]: it checks if bytes are available before calling read

File: libraries/AP_ICEngine/AP_ICEngine.cpp [hunk 3]
```diff
@@ -311,6 +328,76 @@ bool AP_ICEngine::engine_control(float start_control, float cold_start, float he
...
 
+/*
+  Update low throttle limit to ensure steady idle for IC Engines
+  return a new min_throttle value
+*/
+void AP_ICEngine::update_idle_governor(int8_t &min_throttle)
+{
+    const int8_t min_throttle_base = min_throttle;
+
+    // Initialize idle poing to min_throttle on the first run
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_ICEngine/AP_ICEngine.cpp]: fixed

File: libraries/AP_EFI/AP_EFI.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_EFI/AP_EFI.cpp]: fixed

File: libraries/AP_EFI/AP_EFI_Serial_MS.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "AP_EFI.h"
+#include "AP_EFI_Backend.h"
+
+// Engine Status Bitmask
+#define ENG_STATUS_READY             (1<<0)
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_Serial_MS.h]: fixed

File: libraries/AP_EFI/AP_EFI_State.h [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+// Time in milliseconds before we declare the EFI to be "unhealthy"
+#define HEALTHY_LAST_RECEIVED_MS 3000
+
+/***************
+ *
+ * Status enums
+ *
+ ***************/
+
+enum Engine_State {
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_State.h]: fixed

[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_State.h]: fixed

File: libraries/AP_EFI/AP_EFI_State.h [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    uint32_t last_updated_ms;
+
+    // Current overall engine state
+    Engine_State engine_state                         : 2;
+ 
+    // If there is an error that does not fit other error types
+    bool general_error                                : 1;
+
+    // Error/status fields 
+    Crankshaft_Sensor_Status crankshaft_sensor_status : 2;
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_EFI/AP_EFI_State.h]: bit fields removed

**Pull request merged**